clients/src/main/java/org/apache/kafka/common/record/ByteBufferLogInputStream.java

```diff
@@ -62,13 +54,31 @@ public MutableRecordBatch nextBatch() throws IOException {
         batchSlice.limit(batchSize);
         buffer.position(buffer.position() + batchSize);
 
-        if (magic < 0 || magic > RecordBatch.CURRENT_MAGIC_VALUE)
-            throw new CorruptRecordException("Invalid magic found in record: " + magic);
-
         if (magic > RecordBatch.MAGIC_VALUE_V1)
             return new DefaultRecordBatch(batchSlice);
         else
             return new AbstractLegacyRecordBatch.ByteBufferLegacyRecordBatch(batchSlice);
     }
 
+    int nextBatchSize() throws CorruptRecordException {
+        int remaining = buffer.remaining();
+        if (remaining < LOG_OVERHEAD)
+            throw new IllegalStateException("Buffer does not contain batch header");
+
+        int recordSize = buffer.getInt(buffer.position() + SIZE_OFFSET);
+        // V0 has the smallest overhead, stricter checking is done later
+        if (recordSize < LegacyRecord.RECORD_OVERHEAD_V0)
+            throw new CorruptRecordException(String.format("Record size %d is less than the minimum record overhead (%d)",
+                    recordSize, LegacyRecord.RECORD_OVERHEAD_V0));
+        if (recordSize > maxMessageSize)
+            throw new CorruptRecordException(String.format("Record size %d exceeds the largest allowable message size (%d).",
+                    recordSize, maxMessageSize));
+
+        if (remaining >= HEADER_SIZE_UP_TO_MAGIC) {
```

Comment:
Shouldn't we do something If this evaluates to false?

Reply:
When invoked from `nextBatch`, we return null and would do this again when there is sufficient data in the buffer (that was the existing behaviour). For log cleaner, this should always succeed since since we are growing buffer beyond `max.message.bytes`.